astropy/coordinates/matching.py

```diff
@@ -272,9 +277,18 @@ def search_around_3d(coords1, coords2, distlimit, storekdtree="kdtree_3d"):
     kdt1 = _get_cartesian_kdtree(coords1, storekdtree, forceunit=cunit)
     idxs1 = []
     idxs2 = []
-    for i, matches in enumerate(kdt1.query_ball_tree(kdt2, distlimit.to_value(cunit))):
-        idxs1.extend(len(matches) * [i])
-        idxs2.extend(matches)
+
+    if distlimit.isscalar:
+        for i, matches in enumerate(
+            kdt1.query_ball_tree(kdt2, distlimit.to_value(cunit))
+        ):
+            idxs1.extend(len(matches) * [i])
+            idxs2.extend(matches)
+    else:
+        for i, point, distance in zip(range(len(distlimit)), kdt1.data, distlimit):
```

Comment:
If here you write
```
for i, (point, distance) in enumerate(zip(*np.broadcast_arrays(kdt1.data, distlimit))):
   ...
```
then I think this branch will work for the scalar case too (note, I don't have the time to try this, so may have a typo; please check).

Note that even now the code is, to me, rather odd, since it cannot deal with multidimensional coordinates. That can also be done with numpy, by using `np.nditer` rather than `zip`. But perhaps that is better for follow-up, since it is an existing problem.

Reply:
I did not benchmark, but my intuition would be that looping in python on the points in kdt1 would be slower than using `query_ball_tree` (where the loop is done in cython by scipy). The if does not cost much, so if possible, I think keeping the two distinct cases for scalar and non scalar radii would be better.

Reply:
I don't see how it would be possible that the outer iteration here would have measurable performance impact at all, while the code gets considerably simpler, both because of combining (and thus reducing the chance of errors) and the ability to remove the error path.

Reply:
Oops, I did not read very carefully - on second thought, it is possible that having a single distance will allow the scipy code to make things faster.  Easiest of course would be just to test it quickly...

Reply:
```python
import timeit

setup = """
import numpy as np
from astropy.coordinates import SkyCoord
import astropy.units as u

coo1_len = 500
coo2_len = 10000

coo1 = SkyCoord(np.random.rand(coo1_len) * 360, np.random.rand(coo1_len)  * 180  -  90, unit="deg")
coo2 = SkyCoord(np.random.rand(coo2_len) * 360, np.random.rand(coo2_len)  * 180  -  90, unit="deg")
"""

n_iter=500

print(timeit.timeit("coo1.search_around_sky(coo2, seplimit=0.5 * u.deg)", setup=setup, number=n_iter))
print(timeit.timeit("coo2.search_around_sky(coo1, seplimit=[0.5] * coo1_len * u.deg)", setup=setup, number=n_iter))
```
1.031760939004016
12.835172551996948

Looks like there is a factor 12 for this case.

Reply:
OK, thanks so much for doing the test - that clearly settles it!

Reply:
To address your other comment, I'll try to do the multidimensional coordinates case (I did not even think about it, I never use multidimensional coordinates). But I don't have time today.

Reply:
I did not find a good way to support the multidimensional coordinates. If you have a suggestion, (other than `np.nditer` because it did not work) then I'm happy to push it, otherwise we can leave it to the next contributor as you suggested.